### ISSUE TITLE
concurrency: support context propagator SPI

### ISSUE THREAD

**Issue by FroMage** [eclipse/microprofile-concurrency#7]
In order for mp-concurrency to replace [Reactive Contexts](https://reactiverse.io/reactive-contexts/) which is currently used by RESTEasy to provide context propagation, it would be very useful if there could be an SPI to declare custom context propagators.

For example, RxJava 1 and 2 both support a plugin system where it is relatively easy to implement automatic context propagation with a context API such as mp-concurrency and Reactive Contexts.

Supporting an SPI for context propagators is as simple as defining an interface with a single `setup` method, and an entry in `META-INF/services` to declare implementations.

With this in hand, mp-concurrency could provide two optional dependencies that provide implementations that hook into RxJava 1 and 2 in order to use mp-concurrency context propagation to propagate contexts automatically to all RxJava pipelines, rather than require users to implement it themselves or depend on a third-party library.

Copied from https://github.com/eclipse/microprofile-sandbox/issues/19

**Comment by FroMage** [comment]
I'll try to re-explain this issue here in case my original explanation was not clear.

In RESTEasy and Reactive Contexts we take advantage of RxJava's plugin system to make context propagation automatic. For example the following code works out of the box:

```java
@GET
public Single<Hello> hello(){
 return Single.just("Hello")
  .delay(2, TimeUnit.SECONDS)
  .map(string -> string); // you can use your context here
}
```

This works because the `resteasy-rxjava2` module imports the `reactive-contexts-propagators-rxjava2` which implements the propagator SPI to register hooks into RxJava2 so that any callback created in an RxJava2 pipeline will capture the context and restore it in all subsequent listeners.

This is much better than requiring the user to use the context propagation API explicily, especially since context propagation is the better default.

For example, `CompletionStage`, which does not support plugins, requires users to do it explicitly by injecting a `ThreadContext` and calling `ctx.withContextCapture()` on _every_ `CompletionStage` they get.

Does this help?

**Comment by FroMage** [comment]
This has been implemented by #48
